Time related functionalities

gege251 opened the pull request:
- [x] Add awaitSlot, waitNSlots, currentTime, currentSlots contract effect handlers
- [x] Add validity range functions
- [ ] Test on real chain with examples

samuelWilliams99 commented on src/BotPlutusInterface/PreBalance.hs:
```diff
@@ -259,6 +272,20 @@ addSignatories ownPkh privKeys pkhs tx =
+validateRange (Interval _ (UpperBound NegInf _)) = False
+validateRange (Interval (LowerBound (Finite lowerBound) _) (UpperBound (Finite upperBound) _))
+  | lowerBound >= upperBound = False
+  | otherwise = True
```
I believe this is redundant, and caught by the below case

samuelWilliams99 commented on test/Spec/BotPlutusInterface/Contract.hs:
```diff
@@ -685,6 +747,29 @@ useWriter = do
+      contract :: Contract () (Endpoint "SendAda" ()) Text Slot
+      contract = waitNSlots 1
+
+  let (result, state) = runContractPure contract initState
```
```suggestion
      (result, state) = runContractPure contract initState
```

samuelWilliams99 commented on test/Spec/MockContract.hs:
```diff
@@ -258,6 +259,18 @@ runPABEffectPure initState req =
+      mc <* modify @(MockContractState w) (tip %~ incTip)
+
+    incTip TipAtGenesis = Tip 1 (BlockId "00") 0
+    incTip Tip {tipSlot, tipBlockId, tipBlockNo} =
+      Tip
+        { tipSlot = tipSlot + 1
+        , tipBlockId = tipBlockId
+        , tipBlockNo = tipBlockNo
+        }
```
We can use record dot here!
```suggestion
    incTip t@Tip {} = t{tipSlot + 1}
```

gege251 replied:
I'm afraid this doesn't work with the plugin:
> Using the preprocessor, but not the GHC plugin:
>
>    expr{lbl1.lbl2 = val} is equivalent to expr{lbl1 = (expr.lbl1){lbl2 = val}}, performing a nested update.
>    expr{lbl * val} is equivalent to expr{lbl = expr.lbl * val}, where * can be any operator.
>    expr{lbl1.lbl2} is equivalent to expr{lbl1.lbl2 = lbl2}.
https://github.com/ndmitchell/record-dot-preprocessor

gege251 replied:
Also, the Tip is a sum type, so I couldn't even use a conventional record update syntax either (I have a non-exhaustive pattern match error)

samuelWilliams99 replied:
Oh I see, thats a shame, disregard then, my bad!

samuelWilliams99 commented on test/Spec/MockContract.hs:
```diff
@@ -295,7 +310,27 @@ mockCallCommand ShellArgs {cmdName, cmdArgs, cmdOutParser} = do
+mockQueryTip = do
+  state <- get @(MockContractState w)
+
+  let slot = Text.pack $ show $ getSlot $ tipSlot (state ^. tip)
+      blockId = decodeUtf8 $ getBlockId $ tipBlockId (state ^. tip)
+      blockNo = Text.pack $ show $ unBlockNumber $ tipBlockNo (state ^. tip)
```
Maybe unpack the tip before hand to avoid repition
Can use namepuns like
```suggestion
  let Tip {tipSlot, tipBlockId, tipBlockNo} = state ^. tip
      slot = Text.pack $ show $ getSlot tipSlot
      blockId = decodeUtf8 $ getBlockId tipBlockId
      blockNo = Text.pack $ show $ unBlockNumber tipBlockNo
```

gege251 replied:
The problem is the same here, Tip is a sum type, so I realised that the record functions are partial here. Used a good old case statement instead.

gege251 commented on examples/plutus-transfer/src/Cardano/PlutusExample/Transfer.hs:
```diff
@@ -0,0 +1,46 @@
+transfer (TransferParams outputPerTx allPayments) = do
+  tell $ Last $ Just "Contract started"
+  let txs =
+        map (mconcat . map (uncurry Constraints.mustPayToPubKey . first PaymentPubKeyHash)) $
+          group outputPerTx allPayments
```
I think TxConstraint doesn't have a Foldable instance. I'll try to make this somewhat more readable.

gege251 commented on src/BotPlutusInterface/CardanoCLI.hs:
```diff
@@ -333,6 +356,20 @@ mintOpts pabConf buildMode mintingPolicies redeemers mintValue =
         else []
     ]
 
+-- | This function does not check if the range is valid, that
```
Oh, must have accidentally deleted the rest

samuelWilliams99 commented on examples/plutus-transfer/src/Cardano/PlutusExample/Transfer.hs:
```diff
@@ -35,10 +36,12 @@ transfer :: TransferParams -> Contract (Last Text) TransferSchema Text ()
   tell $ Last $ Just "Finished"
+  where
+    toTx :: [(PubKeyHash, Value)] -> TxConstraints Void Void
+    toTx = mconcat . map (uncurry Constraints.mustPayToPubKey . first PaymentPubKeyHash)
```
Theres mconcatMap in Data.List.Extra, but unsure if its worth it. Happy to merge as is

gege251 commented:
@samuelWilliams99 Sorry for adding these after you approved this, but I wanted to incorporate a few more changes in this PR.
- Added the SlotConfig to PABConf, for most cases this can be set to `def`, but might need to be changed for some testnets
- Added support for some other time related requests

samuelWilliams99 commented:
No worries! Looks good to me :)
